openpathsampling/pathmover.py

```diff
@@ -1244,6 +1286,146 @@ def move(self, globalstate):
+        the PathMovers to choose from
+    weights : list of floats
+        the relative weight of each PathMover (does not need to be normalized)
+    """
+
+    def _selector(self, globalstate):
+        if self.weights is None:
+            weights = [1.0] * len(self.movers)
+        else:
+            weights = self.weights
+
+        # this is implemented by setting all weights locally to zero that
+        # correspond to movers that will potentially fail since the required
+        # input ensembles are not present in the globalstate
+
+        present_ensembles = globalstate.ensembles
+
+        for idx, mover in enumerate(self.movers):
+            for ens in mover.input_ensembles:
+                if ens not in present_ensembles:
+                    # ens might be required but is not present
+                    weights[idx] = 0.0
+
+        return weights
+
+class FirstAllowedMover(RandomSelectionMover):
+    """
+    Chooses a first mover that has samples in all required ensembles.
+
+    A mover can only safely be run, if all inputs can be satisfied. This will pick
+    the first mover from the list where all ensembles from input_ensembles are
+    found.
+
+    Attributes
+    ----------
+    movers : list of PathMover
+        the PathMovers to choose from
+    """
+
+    def _selector(self, globalstate):
+        weights = [1.0] * len(self.movers)
+
+        present_ensembles = globalstate.ensembles
+
+        found = False
+
+        for idx, mover in enumerate(self.movers):
+            if not found:
+                for ens in mover.input_ensembles:
+                    if ens not in present_ensembles:
+                        # ens might be required but is not present
+                        weights[idx] = 0.0
+
+                if weights[idx] > 0.0:
+                    found = True
+            else:
+                weights[idx] = 0.0
+
+        return weights
+
+class LastAllowedMover(RandomSelectionMover):
+    """
+    Chooses the last mover that has samples in all required ensembles.
+
+    A mover can only safely be run, if all inputs can be satisfied. This will pick
+    the first mover from the list where all ensembles from input_ensembles are
+    found.
+
+    Attributes
+    ----------
+    movers : list of PathMover
+        the PathMovers to choose from
+    """
+
+    def _selector(self, globalstate):
+        weights = [1.0] * len(self.movers)
+
+        present_ensembles = globalstate.ensembles
+
+        found = False
+
+        for idx, mover in reversed(enumerate(self.movers)):
+            if not found:
+                for ens in mover.input_ensembles:
+                    if ens not in present_ensembles:
+                        # ens might be required but is not present
+                        weights[idx] = 0.0
+
+                if weights[idx] > 0.0:
+                    found = True
+            else:
+                weights[idx] = 0.0
+
+        return weights
```

Comment:
Not sure if I agree on this one. The idea was not to make it look like okeying detailed balance. The idea was to have a clean way to write bootstrapping moves the same way as all the other moves. 
The way I think about moves is that there are some that obey detailed balane for a very good reason while others do not. Also for a good reason. This way the code looked cleaner to me. And we might reuse the `FirstAllowed` and `LastAllowed`.

These two movers fall into the same category as the `SequentialAllowed` mover that is new and was meant for single replica TIS. From a set of possible ones it selects the one that is possible (according to input/output ensembles). Instead of having to have this selection process to program in the PathSimulator.

I agree that this introduces a few more Movers that could be confusing to a user, but I assume that these functions will usually just be ignored and are only for advances users relevant.